### PR TITLE
Moving to localectl for all locale related activities

### PR DESCRIPTION
**Pull request by Torxed**
This should solve a lot of inconsistencies in terms of both listing but also setting locales.
This should also translate well over to X11 (but not wayland), see https://github.com/archlinux/archinstall/issues/107#issuecomment-775262644 for more information.

### CODE REVIEW

**Comment by Torxed** [comment]
One issue remains, and that is `X11` is not available on the ISO, so performing keyboard locale selection based off `localectl --no-pager list-x11-keymap-layouts` won't work. Because:

2021-05-15-180605_1285x99_scrot

**Comment by dylanmtaylor** [comment]
> One issue remains, and that is `X11` is not available on the ISO, so performing keyboard locale selection based off `localectl --no-pager list-x11-keymap-layouts` won't work. Because:
> 
> https://github.com/archlinux/archinstall/blob/b4fcc97bf182ce2406a907dff49ce2dbbdab8a41/archinstall/lib/locale_helpers.py#L13-L15
> 
> 2021-05-15-180605_1285x99_scrot

Can we leverage arch-chroot somehow to run the locale command on the target? It'd be halfway through the install though which is ... not ideal.

**Comment by Torxed** [comment]
> > One issue remains, and that is `X11` is not available on the ISO, so performing keyboard locale selection based off `localectl --no-pager list-x11-keymap-layouts` won't work. Because:
> > https://github.com/archlinux/archinstall/blob/b4fcc97bf182ce2406a907dff49ce2dbbdab8a41/archinstall/lib/locale_helpers.py#L13-L15
> > 
> > 2021-05-15-180605_1285x99_scrot
> 
> Can we leverage arch-chroot somehow to run the locale command on the target? It'd be halfway through the install though which is ... not ideal.

Since we don't need to list the locales, if what @phisch mentioned holds up:

> This somehow sets the keymap for both x11 and vc, BUT from my testing, this only works after the locale has been set to an empty string. So, running localectl set-keymap "" && localectl set-keymap "de".

Then we'd only have to set the locale using `localectl` during installation, and if `x11` packages are installed at the time of executing, both should get set. But this means locale should be set as late as possible to ensure profiles have been installed *(which are responsible for `x11` packages)*.

File: archinstall/lib/locale_helpers.py [hunk 1]
```diff
@@ -26,12 +22,27 @@ def verify_keyboard_layout(layout):
...
 	return False
 
+def verify_x11_keyboard_layout(layout):
+	for language in list_x11_keyboard_languages():
+		if layout.lower() == language.lower():
+			return True
+	return False
 
-def search_keyboard_layout(layout_filter):
+def search_keyboard_layout(layout):
```
Comment (dylanmtaylor) [inline archinstall/lib/locale_helpers.py:31]: would prefer to not undo this change, since linter doesn't like shadowing filter.

File: .gitignore [hunk 2]
```diff
@@ -20,7 +20,7 @@ SAFETY_LOCK
 **/**.network
 **/**.target
 **/**.qcow2
-**/test.py
+**/test*.py
```
Comment (dylanmtaylor) [inline .gitignore:23]: This might bite us if we add pytest tests and happen to use this naming convention.

Comment (Torxed) [reply .gitignore:23]: Good catch!

File: archinstall/install.log [hunk 3]
```diff
@@ -0,0 +1,54 @@
+Not enough permission to place log file at /var/log/archinstall/install.log, creating it in /home/anton/github/archinstall/archinstall/install.log instead.
```
Comment (dylanmtaylor) [inline archinstall/install.log:1]: Drop this file from the PR and add an ignore for it

File: profiles/xorg.py [hunk 4]
```diff
@@ -44,4 +44,4 @@ def _prep_function(*args, **kwargs):
 		else:
 			archinstall.storage['installation_session'].add_additional_packages(f"xorg-server xorg-xinit {' '.join(_gfx_driver_packages)}")
 	except:
-		archinstall.storage['installation_session'].add_additional_packages("xorg-server xorg-xinit")  # Prep didn't run, so there's no driver to install
+		archinstall.storage['installation_session'].add_additional_packages(f"xorg-server xorg-xinit")  # Prep didn't run, so there's no driver to install
```
Comment (dylanmtaylor) [inline profiles/xorg.py:47]: This should not be an f-string.

File: archinstall/lib/installer.py [hunk 5]
```diff
@@ -515,11 +517,38 @@ def user_set_shell(self, user, shell):
...
+			# Setting an empty keymap first, allows the subsequent call to set layout for both console and x11.
+			self.arch_chroot('localectl set-keymap ""')
+
+			if (output := self.arch_chroot(f'localectl set-keymap {language}')).exit_code != 0:
+				raise ServiceException(f"Unable to set locale '{language}' for console: {output}")
+
+			self.log(f"Keyboard language for this installation is now set to: {language}")
 		else:
-			self.log('Keyboard language was not changed from default (no language specified).', fg="yellow", level=logging.INFO)
+			self.log(f'Keyboard language was not changed from default (no language specified).', fg="yellow", level=logging.INFO)
```
Comment (dylanmtaylor) [inline archinstall/lib/installer.py:535]: This also should not be an f-string

**Comment by dylanmtaylor** [comment]
https://github.com/archlinux/archinstall/pull/476 corrects the issues I just pointed out.

File: archinstall/install.log (hunk 3 above)
Comment (Torxed) [reply archinstall/install.log:1]: Thought we had it already, will do :)

**Comment by Torxed** [comment]
Merged those changes. Carrying on with `archinstall.Boot()`

**Comment by Torxed** [comment]
`localectl` now properly sets the `vconsole` configuration as well as `x11` configuration.
So I'm happily merging this in now! There might be some commands that still freak out over the new `SysCommand`, but we'll deal with it then.

screenshot

screenshot
